### PR TITLE
Prepare for public testnet

### LINKED ISSUE #647: Persist fork choice
## Description

Presently, fork choice is not persisted to the database.

## Steps to resolve

- Ensure the beacon chain stores the fork choice when it calls `self.persist`.
- Ensure the beacon chain loads the fork choice back from disk again.

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

Resolves #647

## Proposed Changes

- Adds the ability for the `account_manager` to submit eth1 deposits.
- Adds bug fixes discovered along the way.
- Adds support for [eth2_testnets](https://github.com/eth2-clients/eth2-testnets/).
- Refactors the beacon node config generation in preparation for the testnet.
- Adds the `HeadTracker` struct to track beacon heads.
- Persists `ForkChoice` to disk.

## Notes

@michaelsproul this commit will be of interest to you: https://github.com/sigp/lighthouse/pull/628/commits/ef3cabb60e78f16dc3194a9540ff40d32f49d0ac

### CODE REVIEW

File: eth2/types/src/chain_spec.rs [hunk 1]
```diff
@@ -256,7 +256,6 @@ impl ChainSpec {
             network_id: 2, // lighthouse testnet network id
             boot_nodes,
             eth1_follow_distance: 16,
-            milliseconds_per_slot: 6_000,
```
Comment (michaelsproul) [inline eth2/types/src/chain_spec.rs]: Intentional? Minimal spec uses 6s slots: https://github.com/ethereum/eth2.0-specs/blob/03fb0979485a204890053ae0b5dcbbe06c7f1f5c/configs/minimal.yaml#L59-L60

File: eth2/utils/deposit_contract/src/lib.rs [hunk 2]
```diff
@@ -51,6 +62,6 @@ mod tests {
 
         let data = eth1_tx_data(&deposit).expect("should produce tx data");
 
-        assert_eq!(data.len(), 388, "bytes should be correct length");
+        assert_eq!(data.len(), 420, "bytes should be correct length");
```
Comment (michaelsproul) [inline eth2/utils/deposit_contract/src/lib.rs]: :maple_leaf:

File: beacon_node/beacon_chain/src/eth1_chain.rs [hunk 3]
```diff
@@ -201,35 +215,120 @@ impl<T: EthSpec, S: Store> CachingEth1Backend<T, S> {
...
+            //
+            // Here we choose the latest block in our voting window.
+            blocks
+                .iter()
+                .rev()
+                .skip_while(|eth1_block| eth1_block.timestamp > voting_period_start_seconds)
+                .skip(eth1_follow_distance as usize)
+                .take(1)
+                .collect::<Vec<_>>()
+                .first()
```
[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/eth1_chain.rs]: (replacing the 3 lines from the `take(1)`)

File: account_manager/src/lib.rs [hunk 4]
```diff
@@ -148,3 +270,175 @@ fn make_validators(
...
+                        deposit_value,
+                        account_index,
+                        password,
+                        log,
+                    )
+                    .map(|()| ((), validators))
+                })
+            })
+            .collect()
+            .map(|_| event_loop)
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply account_manager/src/lib.rs]: Very good! Thank you :)

**Comment by paulhauner** [comment]
All comments addressed!

**Pull request merged**